### PR TITLE
SoundManager v0

### PR DESCRIPTION
**Pull request by Vroonsh**
new SoundManager based on ManagerWindow
Base Sounddata exposure & settings (expose sound position in Sceneview based on Balance/Fade)

image

- no import/export yet, as i don't know how we want to handle sounds assets in VPE.
- no playback 

Will partially resolve issue #163

### CODE REVIEW

File: VisualPinball.Unity/VisualPinball.Unity/VPT/Table/TableSerializedContainer.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using VisualPinball.Engine.VPT;
+using VisualPinball.Engine.VPT.Table;
+
+namespace VisualPinball.Unity.VPT.Table
```
Comment (freezy) [inline VisualPinball.Unity/VisualPinball.Unity/VPT/Table/TableSerializedContainer.cs:6]: Change to `VisualPinball.Unity`

File: VisualPinball.Unity/VisualPinball.Unity/VPT/Table/TableSerializedData.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+﻿using UnityEngine;
+using VisualPinball.Engine.VPT;
+
+namespace VisualPinball.Unity.VPT.Table
```
Comment (freezy) [inline VisualPinball.Unity/VisualPinball.Unity/VPT/Table/TableSerializedData.cs:4]: Change to `VisualPinball.Unity`

File: VisualPinball.Unity/VisualPinball.Unity/VPT/Table/TableSerializedSoundContainer.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using VisualPinball.Engine.VPT.Sound;
+using VisualPinball.Engine.VPT.Table;
+
+namespace VisualPinball.Unity.VPT.Table
```
Comment (freezy) [inline VisualPinball.Unity/VisualPinball.Unity/VPT/Table/TableSerializedSoundContainer.cs:7]: Change to `VisualPinball.Unity`

**Review by freezy — CHANGES_REQUESTED** [review]
Thanks! Good first impression, just commented on a few cosmetics. Will test later tonight.

Meanwhile could you explain `TableSerializedContainer`? I must say I haven't completely understood the container concept, so some context would be useful :)

**Comment by Vroonsh** [comment]
> Thanks! Good first impression, just commented on a few cosmetics. Will test later tonight.
> 
> Meanwhile could you explain `TableSerializedContainer`? I must say I haven't completely understood the container concept, so some context would be useful :)

It was formerly created by @rbxnk for Texture on the Image Manager issue (TableSerializedTextureContainer), I just created a generic version of the Container and the serialized data so i can use it for both sound & collections.

File: VisualPinball.Unity/VisualPinball.Unity.Editor/Managers/ImageManager.cs [hunk 4]
```diff
@@ -106,10 +106,7 @@ protected override void OnDataChanged(string undoName, ImageListData data)
 
 		protected override void AddNewData(string undoName, string newName) {
 			_table.Textures.SetNameMapDirty();
-			var sidecar = _table.GetComponentInChildren<TableSidecar>();
-			if (sidecar != null) {
-				Undo.RecordObject(sidecar, undoName);
-			}
+			Undo.RecordObject(_table, undoName);
```
[the remark this answers is not in the document]
Comment (Vroonsh) [reply VisualPinball.Unity/VisualPinball.Unity.Editor/Managers/ImageManager.cs:109]: sidecar is no longer a Component, so the call to GetComponentInChildren was crashing, preventing to add or remove any image

File: VisualPinball.Unity/VisualPinball.Unity.Editor/Managers/SoundListData.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
+﻿using VisualPinball.Engine.VPT;
+
+namespace VisualPinball.Unity.Editor.Managers
```
[the remark this answers is not in the document]
Comment (Vroonsh) [reply VisualPinball.Unity/VisualPinball.Unity.Editor/Managers/SoundListData.cs:3]: I just placed my manager files in the same place as MaterialManager & ImageManager which are in namespace VisualPinball.Unity.Editor.Managers. Do we have to change all namespaces ?

File: VisualPinball.Unity/VisualPinball.Unity/VPT/Table/TableSerializedContainer.cs (hunk 1 above)
Comment (Vroonsh) [reply VisualPinball.Unity/VisualPinball.Unity/VPT/Table/TableSerializedContainer.cs:6]: Same, placed next to existing files, do i have to clean that up ?

**Comment by freezy** [comment]
Awesome animation for the position :)

Any way to make it movable in the scene view?

I tested importing a table, moving sound position around, exporting to VPX, deleting table, and I'm getting this when re-importing the exported VPX:

```
MissingReferenceException: The object of type 'TableAuthoring' has been destroyed but you are still trying to access it.
Your script should either check if it is null or you should not destroy the object.
VisualPinball.Unity.Editor.Managers.SoundManager.OnSceneGUI (UnityEditor.SceneView sceneView) (at C:/Development/VisualPinball.Engine/VisualPinball.Unity/VisualPinball.Unity.Editor/Managers/SoundManager.cs:78)
UnityEditor.SceneView.CallOnSceneGUI () (at <d1bec46880064709a5e713ad543e6d96>:0)
UnityEditor.SceneView.HandleSelectionAndOnSceneGUI () (at <d1bec46880064709a5e713ad543e6d96>:0)
UnityEditor.SceneView.OnGUI () (at <d1bec46880064709a5e713ad543e6d96>:0)
UnityEditor.HostView.InvokeOnGUI (UnityEngine.Rect onGUIPosition, UnityEngine.Rect viewRect) (at <d1bec46880064709a5e713ad543e6d96>:0)
UnityEditor.DockArea.DrawView (UnityEngine.Rect viewRect, UnityEngine.Rect dockAreaRect) (at <d1bec46880064709a5e713ad543e6d96>:0)
UnityEditor.DockArea.OldOnGUI () (at <d1bec46880064709a5e713ad543e6d96>:0)
UnityEngine.UIElements.IMGUIContainer.DoOnGUI (UnityEngine.Event evt, UnityEngine.Matrix4x4 parentTransform, UnityEngine.Rect clippingRect, System.Boolean isComputingLayout, UnityEngine.Rect layoutSize, System.Action onGUIHandler, System.Boolean canAffectFocus) (at <9c9fa79efe9e479a8cbe7a6f99f673d3>:0)
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr, Boolean&)
```

File: VisualPinball.Unity/VisualPinball.Unity.Editor/Managers/ImageManager.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (freezy) [reply VisualPinball.Unity/VisualPinball.Unity.Editor/Managers/ImageManager.cs:109]: Got it, thanks!

File: VisualPinball.Unity/VisualPinball.Unity.Editor/Managers/SoundListData.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (freezy) [reply VisualPinball.Unity/VisualPinball.Unity.Editor/Managers/SoundListData.cs:3]: Leave the existing ones, they are already fixed in another branch.

File: VisualPinball.Unity/VisualPinball.Unity/VPT/Table/TableSerializedContainer.cs (hunk 1 above)
Comment (freezy) [reply VisualPinball.Unity/VisualPinball.Unity/VPT/Table/TableSerializedContainer.cs:6]: Leaving them will make it easier when I merge the other branch :)

**Comment by Vroonsh** [comment]
> Awesome animation for the position :)
> 
> Any way to make it movable in the scene view?
> 

Would try to make it a Handle in a next PR :)

**Comment by freezy** [comment]
@rbxnk could you review the container thing? Otherwise it looks good to me!

**Review by freezy — APPROVED** [review]

**Comment by rbocksnick** [comment]
Container looks good to me. I had intended to do basically the same thing but wanted to wait until we had the use case for it, now we do :-)

For some history on why its there - I wanted a way to store the big chunky vpx data like textures/sounds in some sort of Unity.Object derived thing so we can pass them individually to Undo funcs, since otherwise we'd have to resort to putting the entire (several hundred meg) sidecar on the undo stack. To avoid creating a nasty dependency on Unity in the other parts of the project I can up with the idea of just swapping out the container wholesale after the import finished and we do the unity-side "MakeSerialized". And... since Unity's serialization only works on PODs and Lists that container needed to implement its own dictionary-like interface to keep the rest of the calling code sane.

File: VisualPinball.Unity/VisualPinball.Unity/VPT/Table/TableSerializedSoundContainer.cs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+namespace VisualPinball.Unity
+{
+	/// <summary>
+	/// Implements a serializable Sound container that can replace the engine table container after import
+	/// </summary>
+	[Serializable]
+	public class TableSerializedSoundContainer : TableSerializedContainer<Sound, SoundData, TableSerializedSound>
+	{
+		public override void Add(Sound value)
```
Comment (rbocksnick) [inline VisualPinball.Unity/VisualPinball.Unity/VPT/Table/TableSerializedSoundContainer.cs:15]: Not something to worry about right now, but I wonder if these funcs will end up being the same for all containers with the exception of the create call. We could probably figure out a way to template that too

**Comment by freezy** [comment]
@rbocksnick Got it, thanks for the summary! :)

**Pull request merged**